### DISCUSSION TITLE
Dependencies that require runtime data to be initialised.

### DISCUSSION THREAD

**Discussion by sergdort** [pointfreeco/swift-dependencies#32] [General]
Hey guys! Thanks for your hard work!

I'm researching of ways to migrate from TCA Environment based dependencies to the new solution. And in many places of our app we have services that need runtime data to be usable. For example we've got a `UserConfigService` that needs `UserData` that we get from our server after successful sign in e.g

```swift
final class UserConfigService {
    private let userData: UserData
    
    init(userData: UserData) {}
}
```

Is there a more elegant solution that you would suggest? Because mine looks something like 

```swift
extension UserConfigService: DependencyKey {
    public static var _liveValue: UserConfigService!
    
    public static var liveValue: UserConfigService {
        return _liveValue
    }
}
```
or 

```swift
final class UserConfigService {
    private let userData: UserData?
    
    init() {}

   func update(with userData: UserData) {}
}
```

I've seen in other treads people suggested to use nested `Dependency` wrappers like:

```swift
final class UserConfigService {
    @Dependency(\.userData)
    private var userData
    
    init() {}
}
```

Would love to hear your thoughts on that!

**Comment by mbrandonw** [comment]
Hi @sergdort, using `@Dependency` is definitely fine to use from other dependencies. We plan on writing more documentation about this.

I will also say that the other two styles you have sketched out are not recommended. The first, the one that uses the implicitly unwrapped optional, will not work because upon first access of a dependency it is cached. So you run the risk of caching the `nil` value. And the second can just be annoying to use in practice since you may forget to call `update(with:)` and you have to deal with an optional.

**Comment by sergdort** [comment]
Thanks, @mbrandonw ! Would it be correct to set a service in a runtime via `DependencyValues` then?

```swift
DependencyValues.live.userSessionServices = UserSessionService(userData: userData)
```

**Reply by mbrandonw** [reply]
Hi @sergdort, that actually isn't possible at all due to the fact that dependencies are stored in a `@TaskLocal`, which does not allow for changing values globally, but rather just for a well-defined scope (read [this article](https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/lifetimes) for more information).

Instead you should just reference `@Dependency(\.userData)` directly in the `UserConfigService`.

For example, [this](https://github.com/pointfreeco/pointfreeco/blob/cc4fbeec6099247db0773a8fa1cea53ea963ce31/Sources/PointFree/Environment.swift#L75-L96) is how we do it in www.pointfree.co to allow the database client to pull environment variables from the dependency.

**Reply by sergdort** [reply]
Thanks for the links! Reading the documentation it seems to me that [`reducer.dependecy`](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/reducerprotocol/dependency(_:_:)) is the method that would suit our needs. 

However I was testing some behaviours and I'm wondering if I have wrong assumption about the [Dependency lifetime](https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/lifetimes/)

```swift
@MainActor
final class ReducerTests: XCTestCase {
    func testDependencyToChildReducers() async {
        let store = TestStore(
            initialState: TestReducer.State(),
            reducer: TestReducer()
                .dependency(\.intDependency, 1)
        )
        
        let store2 = TestStore(
            initialState: TestReducer.State(),
            reducer: TestReducer()
        )
        
        await store.send(.child(.increment)) {
            $0.child.count = 3
        }
        .finish()
        
        await store2.send(.child(.increment)) {
            $0.child.count = 44 // Here I'm expecting the value to be 85 like in the test bellow
        }
        .finish()
    }
    
    func testDependencyToChildReducers2() async {
        let store = TestStore(
            initialState: TestReducer.State(),
            reducer: TestReducer()
        )
        
        await store.send(.child(.increment)) {
            $0.child.count = 85
        }
        .finish()
    }
    
    struct TestReducer: ReducerProtocol {
        @Dependency(\.intDependency)
        var intValue
        
        struct State: Equatable {
            var count = 0
            var child = ChildReducer.State()
        }

        enum Action: Equatable {
            case increment
            case decrement
            case child(ChildReducer.Action)
        }
        
        var body: Reduce<State, Action> {
            Scope(state: \.child, action: /Action.child) {
                ChildReducer()
            }
            Reduce { state, action in
                switch action {
                case .increment:
                    state.count += intValue
                case .decrement:
                    state.count += intValue
                case .child:
                    break
                }
                return .none
            }
        }
    }
    
    
    struct ChildReducer: ReducerProtocol {
        @Dependency(\.intDependency)
        var intValue
        @Dependency(\.intDependency2)
        var intValue2
        
        struct State: Equatable {
            var count = 0
        }

        enum Action: Equatable {
            case increment
            case decrement
        }
        
        func reduce(into state: inout State, action: Action) -> EffectTask<Action> {
            switch action {
            case .increment:
                state.count += (intValue + intValue2)
            case .decrement:
                state.count -= (intValue + intValue2)
            }
            return .none
        }
    }
    
    enum IntDependencyKey: DependencyKey {
        static let liveValue: Int = 42
        static let testValue: Int = 42
    }
    
    enum IntDependency2Key: DependencyKey {
        static var liveValue: Int {
            @Dependency(\.intDependency)
            var intDependency
            
            return intDependency + 1
        }
        static var testValue: Int {
            @Dependency(\.intDependency)
            var intDependency
            
            return intDependency + 1
        }
    }
}

extension DependencyValues {
    var intDependency: Int {
        set {
            self[ReducerTests.IntDependencyKey.self] = newValue
        }
        get {
            self[ReducerTests.IntDependencyKey.self]
        }
    }
    
    var intDependency2: Int {
        set {
            self[ReducerTests.IntDependency2Key.self] = newValue
        }
        get {
            self[ReducerTests.IntDependency2Key.self]
        }
    }
}

```

**Reply by mbrandonw** [reply]
Hey @sergdort, thanks for the detailed test case, that was really helpful.

It may seem confusing, but this is actually the correct behavior and we just need more documentation on this facet. We don't think people are going to run into it often because it is only relevant when running multiple "single entry point systems" at the same time. In this case, your two `TestStore`s.

In the background, when a dependency is first accessed its value is cached. [Here's][cache-pollution-test] a test that shows that. Further, two `TestStore`s created in the same same context are going to share a cache. This behavior is intended because it is what would happen in a real application if you constructed two `Store`s.

We will add some documentation to this caching behavior soon.

[cache-pollution-test]: https://github.com/pointfreeco/swift-dependencies/blob/5493b96bcd573656edd32df8fb802dcb4b9ea678/Tests/DependenciesTests/DependencyValuesTests.swift#L582-L592

**Reply by sergdort** [reply]
Thanks! I see, so dependencies do not share the "Store" scope, but have a global resolution mechanism 👍

**Reply by mbrandonw** [reply]
I'm not sure what you mean by _share the "Store" scope_, but you are correct that by default all access to `@Dependency` goes through a "global" cache mechanism. I put global in quotes because it is very easy to detach from that cache to have a completely new cache:

```swift
withDependencies {
  $0 = DependencyValues()
} operation: {
  FeatureModel()
}
```

That will cause the `FeatureModel` to run in a context where no dependencies are cached and everything is reset back to its defaults.

**Reply by sergdort** [reply]
I guess by `Store` scope I mean that if we apply `dependency` method to a reducer e.g

```swift
let store = Store(initialState: state, reducer: MyReducer().dependency(\.myDependency, dependency))
```
Then `MyReducer` and any its children will inherit the value of the `dependency`, while it does not seem to be the case. Like in my mind I was thinking that once we do that a new instance of `DependencyValues` created (with a new cache instance etc) which is associated with this `Store` instance.

**Reply by mbrandonw** [reply]
I think the reason that things are that way is because typically you only create one single store at the root of the application, and then derive new stores via `store.scope`. In that set up it doesn't matter if you clear the cache or not.

However, we do want to explore more into what it means to have multiple, disconnected stores in an application, and maybe when we get deep into that we will change our mind about the behavior. Maybe whenever you construct a store directly (as opposed to via `scope`) it should clear the dependency cache automatically. Or maybe there's some kind of `Store.detach` operator (just spitballing).

But until then you can accomplish what you want like this:

```swift 
Store(
  initialState: …
  reducer: AppReducer()
    .dependency(\.self, DependencyValues())
    .dependency(…)
```

That will clear the cache for just this one store.

And once we make the newest TCA release (0.50) you will just be able to supply a trailing closure to `Store` to customize the dependencies:

```swift
Store(…) {
  $0 = DependencyValues()
  $0.myDependency = …
}
```

**Reply by sergdort** [reply]
Thanks @mbrandonw. 
We do use multiple stores in our app as we had some performance issues and Swift limitations. But yeah even with this setup we don't want these stores to use different dependencies at runtime. 
It was more like an exercise for me to figure out the Dependencies lifetime =)
